src/Lumper.Lib/BSP/Lumps/BspLumps/PakFileLump.cs

```diff
@@ -9,132 +10,201 @@ namespace Lumper.Lib.BSP.Lumps.BspLumps;
+/// </summary>
+public class PakfileLump(BspFile parent) : ManagedLump<BspLumpType>(parent), IFileBackedLump
 {
     public List<PakfileEntry> Entries { get; private set; } = [];
 
-    private ZipArchive _zip;
     [JsonIgnore]
-    public ZipArchive Zip
-    {
-        get => _zip;
-        set
-        {
-            _zip = value;
-            UpdateEntries();
-        }
-    }
+    public Stream DataStream { get; set; } = null!;
+
+    public long DataStreamOffset { get; set; }
+
+    public long DataStreamLength { get; set; }
+
+    [JsonIgnore]
+    public ZipArchive Zip { get; private set; } = null!;
 
-    public override void Read(BinaryReader reader, long length)
+    private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
+
+    // Empty methods to avoid OO spaghetti, must call our Read/Write methods instead.
+    public override void Read(BinaryReader reader, long length) => throw new NotImplementedException();
+
+    public override void Write(Stream stream) => throw new NotImplementedException();
+
+    public void Read(BinaryReader reader, long length, IoHandler handler)
     {
+        DataStream = reader.BaseStream;
+        DataStreamOffset = reader.BaseStream.Position;
+        DataStreamLength = length;
+
         Stream stream = reader.BaseStream;
         var dataStream = new MemoryStream();
         int read;
-        var buffer = new byte[80 * 1024];
+        const int bufferSize = 80 * 1024;
+        var incr = (float)bufferSize / length * (int)IoHandler.ReadProgressProportions.Paklump;
+        var buffer = new byte[bufferSize];
 
-        var remaining = length;
-        while ((read = stream.Read(buffer, 0, (int)Math.Min(buffer.Length, remaining))) > 0)
+        handler.UpdateProgress(0, "Reading pakfile");
+        var remaining = (int)length;
+        while ((read = stream.Read(buffer, 0, int.Min(bufferSize, remaining))) > 0)
         {
+            if (handler.Cancelled)
+                return;
+
             dataStream.Write(buffer, 0, read);
             remaining -= read;
+            handler.UpdateProgress(incr);
         }
 
         dataStream.Seek(0, SeekOrigin.Begin);
         Zip = ZipArchive.Open(dataStream);
-    }
+        Entries = Zip.Entries.Select(entry => new PakfileEntry(entry)).ToList();
 
-    private void UpdateEntries()
-    {
-        Entries.Clear();
-        foreach (ZipArchiveEntry entry in Zip.Entries)
-        {
-            Entries.Add(new PakFileEntry(entry));
-        }
+        // Mark the zip as compressed if we have any LZMA-compressed entries. We won't update an existing
+        // entry when the lump has ShouldCompress: DesiredCompression.DontChange, but if a new item is added
+        // we'll use this value.
+        IsCompressed = Zip.Entries.Any(e => e is { CompressionType: CompressionType.LZMA });
     }
 
-    private void UpdateZip(bool closeStream)
+    public void Write(Stream stream, IoHandler handler, DesiredCompression compression)
```

Comment:
a bunch of early returns, missing usings on streams and maybe the comment about the Read overload applies to this too. Could maybe split the function up a bit to get rid of the returns and make it more readable but I haven't looked at it too closely yet.

Reply:
Alright, added some `else if`/`else`s, not that I think the early returns are that bad here, doesn't take much to get used to a `return` at the end of an `if` block.

Have done the `using`s as well but don't have time for rebasing right now